src/src/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -595,33 +589,31 @@ private AuthenticationResult processTokenResponse(HttpWebResponse webResponse) {
                 correlationIdInHeader = listOfHeaders.get(0);
             }
         }
-
-        if (webResponse.getStatusCode() == HttpURLConnection.HTTP_OK
-                && webResponse.getBody() != null && webResponse.getBody().length > 0) {
-            // invalid refresh token calls has error related items in the body.
-            // Status is 400 for those.
-            try {
-                String jsonStr = new String(webResponse.getBody());
-                extractJsonObjects(responseItems, jsonStr);
-                result = processUIResponseParams(responseItems);
-            } catch (final JSONException ex) {
-                // There is no recovery possible here, so
-                // catch the
-                // generic Exception
-                Logger.e(TAG, ex.getMessage(), "", ADALError.SERVER_INVALID_JSON_RESPONSE, ex);
-                result = new AuthenticationResult(JSON_PARSING_ERROR, ex.getMessage(), null);
-            }
+        
+        final boolean isResponseBodyEmpty = (webResponse.getBody() == null) || (webResponse.getBody().length == 0);
+        final int statusCode = webResponse.getStatusCode();
+        if (isResponseBodyEmpty) {
+            // Empty response body, return the status code as error code back
+            Logger.v(TAG, "Token request returned an empty response body.");
+            result = new AuthenticationResult(String.valueOf(statusCode), "", null);
         } else {
-            String errMessage = null;
-            byte[] message = webResponse.getBody();
-            if (message != null) {
-                errMessage = new String(message);
-            } else {
-                errMessage = "Status code:" + String.valueOf(webResponse.getStatusCode());
+            switch(statusCode) {
+            case HttpURLConnection.HTTP_OK:
+            case HttpURLConnection.HTTP_BAD_REQUEST:
+            case HttpURLConnection.HTTP_UNAUTHORIZED: 
+                try {
+                    result = parseJsonResponse(webResponse.getBody());
+                } catch (final JSONException ex) {
+                    // Cannot parse response as JSON, then return the raw response as error description
+                    final String message = "Status code: " + statusCode + ". Cannot parse response as JSON " 
+                            + ex.getMessage();
+                    Logger.e(TAG, message, "", ADALError.SERVER_INVALID_JSON_RESPONSE, ex);
+                    result = new AuthenticationResult(String.valueOf(statusCode), webResponse.getBody().toString(), null);
```

Comment:
Did you mean `new String(webResponse.getBody())` instead of `webResponse.getBody().toString()`?